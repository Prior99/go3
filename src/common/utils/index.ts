export * from "./board-color";
export * from "./board-sizes";
export * from "./encrypt";
export * from "./environment";
export * from "./is-browser";
export * from "./rating";
export * from "./rank";
export * from "./authorization";
