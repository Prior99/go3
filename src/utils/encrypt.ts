import { createHmac } from "crypto";

const SECRET = "Eech3ahm";

export function hash(value: string) {
    return createHmac("sha256", SECRET).update(value).digest("hex");
}
