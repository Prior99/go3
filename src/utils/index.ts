export * from "./environment";
export * from "./require-login";
