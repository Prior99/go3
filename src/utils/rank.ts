export enum RankClass {
    KYU = "kyu",
    DAN = "dan",
}

export class Rank {
    public rank: number;
    public rankClass: RankClass;
    public rating: number;

    constructor(rating: number) {
        this.rating = rating;
        if (rating < 2100) {
            this.rank = Math.round((2000 - rating) / 100 + 1);
            this.rankClass = RankClass.KYU;
        } else {
            this.rank = Math.round((rating - 2000) / 100);
            this.rankClass = RankClass.DAN;
        }
    }

    public format() {
        if (this.rankClass === RankClass.KYU) {
            return `${this.rank} Kyu`;
        }
        return `${this.rank} Dan`;
    }

    public formatShort() {
        if (this.rankClass === RankClass.KYU) {
            return `${this.rank}k`;
        }
        return `${this.rank}d`;
    }

    public equals(other: Rank) {
        return this.rank === other.rank && this.rankClass === other.rankClass;
    }

    public greaterThan(other: Rank) {
        if (this.rankClass === other.rankClass) {
            return this.rankClass === RankClass.KYU ? this.rank < other.rank : this.rank > other.rank;
        }
        if (this.rankClass === RankClass.DAN) {
            return true;
        }
        return false;
    }
}

export function formatRank(rating: number) {
    const rank = new Rank(rating);
    return rank.format();
}

export function formatRankShort(rating: number) {
    const rank = new Rank(rating);
    return rank.formatShort();
}
