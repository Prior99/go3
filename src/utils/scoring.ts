export enum GameResult {
    WIN = "win",
    LOSS = "loss",
    TIE = "tie",
}

export enum RankClass {
    KYU = "kyu",
    DAN = "dan",
}

export interface Rank {
    rank: number;
    rankClass: RankClass;
}

/**
 * Implementation of the official EGF rating system.
 * http://www.europeangodatabase.eu/EGD/EGF_rating_system.php
 */
const epsilon = 0.016;

function winningExpectancy(ratingDifference: number) {
    // SE(A) = 1 / (e ^ (D / a) + 1) - epsilon / 2
    return 1 / (Math.exp(ratingDifference / adjustment(ratingDifference)) + 1) - epsilon / 2;
}

function adjustment(rating: number) {
    return 205 - Math.floor(rating / 20);
}

function magnitude(rating: number) {
    return Math.round(122 - 6 * (rating / 100) + Math.pow(rating / 100, 2) / 15);
}

function scoreFromResult(result: GameResult) {
    switch (result) {
        case GameResult.WIN: return 1;
        case GameResult.TIE: return 0.5;
        case GameResult.LOSS: return 0;
        default: return;
    }
}

export function newRating(oldRating: number, otherRating: number, result: GameResult) {
    const actualScore = scoreFromResult(result);
    const expectedScore = winningExpectancy(otherRating - oldRating);
    let difference = actualScore - expectedScore;
    return Math.round(oldRating + magnitude(oldRating) * (actualScore - expectedScore));
}

export function rankFromRating(rating: number): Rank {
    if (rating < 2100) {
        return {
            rank: Math.round((2000 - rating) / 100 + 1),
            rankClass: RankClass.KYU,
        };
    }
    return {
        rank: Math.round((rating - 2000) / 100),
        rankClass: RankClass.DAN,
    };
}

export function formatRank(rating: number) {
    const rank = rankFromRating(rating);
    if (rank.rankClass === RankClass.KYU) {
        return `${rank.rank} Kyu`;
    }
    return `${rank.rank} Dan`;
}

export function formatRankShort(rating: number) {
    const rank = rankFromRating(rating);
    if (rank.rankClass === RankClass.KYU) {
        return `${rank.rank}k`;
    }
    return `${rank.rank}d`;
}
