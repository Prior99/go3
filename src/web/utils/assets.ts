import { component } from "tsdi";
import { bindAll } from "lodash-decorators";
import * as downscale from "downscale";

function getKey(url: string, width?: number, height?: number) {
    if (typeof width !== "undefined" && typeof height !== "undefined") {
        return `${width}__${height}__${url}`;
    }
    return url;
}

@component
@bindAll()
export class Assets {
    private images = new Map<string, HTMLImageElement>();
    private loading = new Map<string, ((HTMLImageElement) => void)[]>();
    private scaling = new Map<string, ((HTMLImageElement) => void)[]>();

    public get loaded() { return this.loading.size === 0; }

    public loadImage(url: string, width?: number, height?: number) {
        if (typeof width === "undefined" || typeof height === "undefined") {
            return this.downloadImage(url);
        }
        return new Promise<HTMLImageElement>(async resolve => {
            const key = getKey(url, width, height);
            if (this.images.has(key)) {
                return resolve(this.images.get(key));
            }
            if (!this.scaling.has(key)) {
                this.scaling.set(key, []);
                const original = await this.downloadImage(url);
                const scaled = await downscale(original, width, height, { imageType: "png" });
                const scaledImg = new Image();
                scaledImg.addEventListener("load", () => {
                    this.images.set(key, scaledImg);
                    this.scaling.get(key).forEach(callback => callback(scaledImg));
                    this.scaling.delete(key);
                });
                scaledImg.src = scaled;
            }
            this.scaling.get(key).push(resolve);
        });
    }

    private downloadImage(url: string) {
        return new Promise<HTMLImageElement>(resolve => {
            if (this.images.has(url)) {
                return resolve(this.images.get(url));
            }
            if (!this.loading.has(url)) {
                this.loading.set(url, []);
                const img = new Image();
                img.addEventListener("load", () => {
                    this.images.set(url, img);
                    this.loading.get(url).forEach(callback => callback(img));
                    this.loading.delete(url);
                });
                img.src = url;
            }
            this.loading.get(url).push(resolve);
        });
    }

    public get(url: string, width?: number, height?: number) {
        return this.images.get(getKey(url, width, height));
    }
}
