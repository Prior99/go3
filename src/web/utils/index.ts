export * from "./assets";
export * from "./rendering";
