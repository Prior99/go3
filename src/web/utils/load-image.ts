import { component } from "tsdi";
import { bind } from "decko";

@component
export class Assets {
    private images = new Map<string, HTMLImageElement>();
    private loading = new Map<string, ((HTMLImageElement) => void)[]>;

    public get loaded() { return loading.length === 0; }

    @bind public loadImage(url: string) {
        return new Promise<HTMLImageElement>(resolve => {
            if (!this.loading.has(url)) {
                this.loading.set(url, []);
            }
            this.loading.get(url).push(resolve);
            const img = new Image();
            img.src = url;
            img.addEventListener("load", () => {
                this.images.set(url, img);
                this.loading.get(url).forEach(callback => callback(img));
                this.loading.delete(url);
            });
        });
    }

    @bind public get(url: string) {
        return this.images.get(url);
    }
}
