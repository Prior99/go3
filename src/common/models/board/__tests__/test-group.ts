import { testBoard } from "./utils";
import { Group } from "..";

test("calculating the freedoms", () => {
    const board = testBoard(
        " B  " +
        " BWB" +
        " WWB" +
        "    ",
    );
    expect(board.groupAt(board.toIndex({ col: 2, row: 1 })).freedoms).toEqual(4);
});

test("calculating the neighbours", () => {
    const board = testBoard(
        " B  " +
        " BWB" +
        " WWB" +
        "    ",
    );
    expect(board.groupAt(board.toIndex({ col: 2, row: 1 })).neighbours.map(board.toPos)).toMatchSnapshot();
});

test("calculating the neighbouring groups", () => {
    const board = testBoard(
        " B  " +
        " BWB" +
        " WWB" +
        "    ",
    );
    expect(
        board
            .groupAt(board.toIndex({ col: 2, row: 1 }))
            .neighbouringGroups.map(group => group.indices.map(board.toPos)),
    ).toMatchSnapshot();
});

test("the size of a group", () => {
    const board = testBoard(
        " B  " +
        " BWB" +
        " WWB" +
        "    ",
    );
    expect(board.groupAt(board.toIndex({ col: 2, row: 1 })).size).toEqual(3);
});

test("checking if a group is the eye of another group", () => {
    const board = testBoard(
        " B B" +
        " BBB" +
        " W  " +
        "    ",
    );
    const group = board.groupAt(board.toIndex({ col: 1, row: 0 }));
    const eye = board.groupAt(board.toIndex({ col: 2, row: 0 }));
    expect(eye.isEyeOf(group)).toBe(true);
});

test("checking if a group is equal to another group", () => {
    const board = testBoard(
        " B B" +
        " BBB" +
        " W  " +
        " WW ",
    );
    const group1a = board.groupAt(board.toIndex({ col: 1, row: 0 }));
    const group1b = board.groupAt(board.toIndex({ col: 3, row: 0 }));
    const group1c = board.groupAt(board.toIndex({ col: 1, row: 1 }));
    const group2 = board.groupAt(board.toIndex({ col: 1, row: 3 }));
    expect(group1a.equals(group1b)).toBe(true);
    expect(group1b.equals(group1c)).toBe(true);
    expect(group1c.equals(group1a)).toBe(true);
    expect(group1a.equals(group2)).toBe(false);
    expect(group1b.equals(group2)).toBe(false);
    expect(group1c.equals(group2)).toBe(false);
});

test("getting the eyes of a group", () => {
    const board = testBoard(
        "WWWW" +
        "W W " +
        "WWWW" +
        "  W ",
    );
    expect(
        board
            .groupAt(board.toIndex({ col: 0, row: 0 })).eyes
            .map(group => group.indices.map(board.toPos)),
    ).toMatchSnapshot();
});
