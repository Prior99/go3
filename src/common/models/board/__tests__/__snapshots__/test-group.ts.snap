// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculating the neighbouring groups 1`] = `
Array [
  Array [
    Object {
      "col": 3,
      "row": 1,
    },
    Object {
      "col": 3,
      "row": 2,
    },
  ],
  Array [
    Object {
      "col": 1,
      "row": 1,
    },
    Object {
      "col": 1,
      "row": 0,
    },
  ],
  Array [
    Object {
      "col": 2,
      "row": 0,
    },
    Object {
      "col": 3,
      "row": 0,
    },
  ],
  Array [
    Object {
      "col": 2,
      "row": 3,
    },
    Object {
      "col": 1,
      "row": 3,
    },
    Object {
      "col": 0,
      "row": 3,
    },
    Object {
      "col": 0,
      "row": 2,
    },
    Object {
      "col": 0,
      "row": 1,
    },
    Object {
      "col": 0,
      "row": 0,
    },
    Object {
      "col": 3,
      "row": 3,
    },
  ],
]
`;

exports[`calculating the neighbours 1`] = `
Array [
  Object {
    "col": 3,
    "row": 1,
  },
  Object {
    "col": 2,
    "row": 2,
  },
  Object {
    "col": 1,
    "row": 1,
  },
  Object {
    "col": 2,
    "row": 0,
  },
  Object {
    "col": 3,
    "row": 2,
  },
  Object {
    "col": 2,
    "row": 3,
  },
  Object {
    "col": 1,
    "row": 2,
  },
  Object {
    "col": 2,
    "row": 1,
  },
  Object {
    "col": 1,
    "row": 3,
  },
  Object {
    "col": 0,
    "row": 2,
  },
]
`;

exports[`getting the eyes of a group 1`] = `
Array [
  Array [
    Object {
      "col": 1,
      "row": 1,
    },
  ],
  Array [
    Object {
      "col": 0,
      "row": 3,
    },
    Object {
      "col": 1,
      "row": 3,
    },
  ],
  Array [
    Object {
      "col": 3,
      "row": 1,
    },
  ],
  Array [
    Object {
      "col": 3,
      "row": 3,
    },
  ],
]
`;
